rasa/core/featurizers.py

```diff
@@ -286,65 +625,79 @@ def _create_states(
+            intent_rows_to_fill = np.where(feature[:, 3] != -1)[0]
+            user_text_rows_to_fill = np.where(feature[:, 3] != 1)[0]
+            action_names_rows_to_fill = np.where(feature[:, 7] != -1)[0]
+            action_text_rows_to_fill = np.where(feature[:, 7] != 1)[0]
+
+            for key in shapes.keys():
+                if INTENT in key:
+                    feature[np.array(intent_rows_to_fill), 2] = [
+                        np.ones((1, shapes.get(key))) * -1
+                    ] * len(intent_rows_to_fill)
+                elif ACTION_NAME in key:
+                    feature[np.array(action_names_rows_to_fill), 6] = [
+                        np.ones((1, shapes.get(key))) * -1
+                    ] * len(action_names_rows_to_fill)
+                elif ACTION_TEXT in key:
+                    if "sparse" in key:
+                        feature[np.array(action_text_rows_to_fill), 4] = [
+                            scipy.sparse.coo_matrix((1, shapes.get(key)))
+                        ] * len(action_text_rows_to_fill)
+                    elif "dense" in key:
+                        feature[np.array(action_text_rows_to_fill), 5] = [
+                            np.ones((1, shapes.get(key))) * -1
+                        ] * len(action_text_rows_to_fill)
+                else:
+                    if "sparse" in key:
+                        feature[np.array(user_text_rows_to_fill), 0] = [
+                            scipy.sparse.coo_matrix((1, shapes.get(key)))
+                        ] * len(user_text_rows_to_fill)
+                    elif "dense" in key:
+                        feature[np.array(user_text_rows_to_fill), 1] = [
+                            np.ones((1, shapes.get(key))) * -1
+                        ] * len(user_text_rows_to_fill)
+
+        return features
+
     def _featurize_states(
-        self, trackers_as_states: List[List[Dict[Text, float]]]
+        self,
+        trackers_as_states: List[List[Dict[Text, float]]],
+        interpreter: NaturalLanguageInterpreter,
     ) -> Tuple[np.ndarray, List[int]]:
         """Create X."""
 
         features = []
         true_lengths = []
 
         for tracker_states in trackers_as_states:
-            dialogue_len = len(tracker_states)
 
             # len(trackers_as_states) = 1 means
             # it is called during prediction or we have
             # only one story, so no padding is needed
 
-            if len(trackers_as_states) > 1:
-                tracker_states = self._pad_states(tracker_states)
-
             story_features = [
-                self.state_featurizer.encode(state) for state in tracker_states
+                self.state_featurizer.encode(state, interpreter)
+                for state in tracker_states
             ]
+            dialogue_len = len(story_features)
+
+            if not story_features == []:
+                features.append(np.stack(story_features))
+                true_lengths.append(dialogue_len)
```

Comment:
do we use true length anywhere? I think with DIET system, we calculate true lengths later

Reply:
we use them in creating the mask for the dialogue; it is useful because when the `features` are stored in training data, each feature is already padded.